pkg/whep/server.go

```diff
@@ -0,0 +1,120 @@
+package whep
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"hash/crc32"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/livekit/protocol/logger"
+	"github.com/livekit/protocol/rpc"
+	"github.com/livekit/psrpc"
+)
+
+const (
+	authorizationHeader = "Authorization"
+	bearerPrefix        = "Bearer "
+	accessTokenParam    = "access_token"
+)
+
+type Server struct {
+	*http.ServeMux
+
+	psrpcClient rpc.WHEPClient
+	logger      logger.Logger
+}
+
+func NewServer(psrpcClient rpc.WHEPClient) *Server {
+	s := &Server{
+		psrpcClient: psrpcClient,
+		logger:      logger.GetLogger(),
+	}
+	r := http.NewServeMux()
+	r.HandleFunc("POST /{room}/{participant}", s.handleNewSession)
+
+	r.HandleFunc("PATCH /{room}/{participant}/{resource_id}", s.handleICE)
+	r.HandleFunc("DELETE /{room}/{participant}/{resource_id}", s.handleDeleteSession)
+	s.ServeMux = r
+
+	return s
+}
+
+func (s *Server) handleNewSession(w http.ResponseWriter, r *http.Request) {
+	authHeader := r.Header.Get(authorizationHeader)
+	var authToken string
+
+	if authHeader != "" {
+		// the request has already passed authorize middleware so this should not happen
+		if !strings.HasPrefix(authHeader, bearerPrefix) {
+			w.WriteHeader(http.StatusUnauthorized)
+			return
+		}
+		authToken = authHeader[len(bearerPrefix):]
+	} else {
+		authToken = r.FormValue(accessTokenParam)
+	}
+
+	room := r.PathValue("room")
+	participant := r.PathValue("participant")
+
+	offer := bytes.Buffer{}
+
+	_, err := io.Copy(&offer, r.Body)
+	if err != nil {
+		s.handleError(w, r, err)
+		return
+	}
+
+	wsUrl := r.FormValue("livekit_url")
```

Comment:
The ws url should use the same host with the whep request, but the request could be rewritten by if the server is behind a proxy.  Is there a way to get the URL (host) from whep request?

Reply:
probably from the http host header?

Reply:
not sure if it could be rewritten too, generate url from both host header and query parameter.